### ISSUE TITLE
GraphTypeTypeRegistry.Register design flaw

### ISSUE THREAD

**Issue by sungam3r** [graphql-dotnet/graphql-dotnet#2189]
I ran into one error while working on #2102. One of the tests started to fail - auto_register_object_graph_type but only as part of the rest of the tests. A separate launch works fine. In the end, I found out the reason. It's funny, but it turned out that this is one of the tasks I had postponed, which was discovered long ago. See `auto_register_object_graph_type` and `auto_register_input_object_graph_type`. The first registers `Money` as `AutoRegisteringObjectGraphType`, the second does the same but as `AutoRegisteringInputObjectGraphType`. `GraphTypeTypeRegistry.Register` rewrites stored value. When tests are run in parallel, the state changes, which leads to an error. But the problem seems to me wider. Now the design of the library does not provide for the registration of the same type at the same time as input and output.

**Comment by sungam3r** [comment]
I mean this place:
```c#
 throw new ArgumentOutOfRangeException(nameof(fieldType),
                            $"Input type '{Name ?? GetType().GetFriendlyName()}' can have fields only of input types: ScalarGraphType, EnumerationGraphType or IInputObjectGraphType. Field '{fieldType.Name}' has an output type.");
```

and that

```c#
     throw new ArgumentOutOfRangeException(nameof(fieldType),
                            $"Output type '{Name ?? GetType().GetFriendlyName()}' can have fields only of output types: ScalarGraphType, ObjectGraphType, InterfaceGraphType, UnionGraphType or EnumerationGraphType. Field '{fieldType.Name}' has an input type.");          
```

**Comment by sungam3r** [comment]
Fixed.
